File: Medium/Greedy/56_Merge_Intervals.cpp
/*
Problem: Merge Intervals

Approach:
- Sort all intervals based on starting time
- Initialize result with first interval
- Traverse remaining intervals:
    - If current interval overlaps with last merged interval:
        merge by updating end = max(end, current_end)
    - Else:
        add current interval to result

Time Complexity: O(n log n)
Space Complexity: O(n)
*/

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        // Sort intervals based on starting time
        sort(intervals.begin(), intervals.end());

        vector<vector<int>> result;
        // Start with the first interval
        result.push_back(intervals[0]);

        // Traverse remaining intervals
        for (int i = 1; i < intervals.size(); i++) {
            // Get last interval in result
            vector<int>& last = result.back();

            // If overlapping
            if (intervals[i][0] <= last[1]) {
                last[1] = max(last[1], intervals[i][1]);
            } else {
                result.push_back(intervals[i]);
            }
        }
        return result;
    }
};

File: Medium/Arrays_TwoPointers/56_Merge_Intervals.py
"""
Problem: Merge Intervals

Approach:
- Sort intervals based on their start time
- Initialize the result with the first interval
- For each remaining interval:
    - If it overlaps with the last interval in the result, merge them
    - Otherwise, add it as a new interval
- Return the merged intervals

Time Complexity: O(n log n)
Space Complexity: O(n)
"""

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        result = [intervals[0]]

        for i in range(1,len(intervals)):
            if(result[-1][1] >= intervals[i][0]):
                result[-1][1] = max(result[-1][1],intervals[i][1])
            else:
                result.append(intervals[i])
        return result
